### PR TITLE
Features/matsa reframe confile

### PR DESCRIPTION
**Pull request by MarsLviv**
reframed class StoreConf

### CODE REVIEW

File: src/settings/storeconf.cpp [hunk 1]
```diff
@@ -9,126 +9,165 @@
...
+{
+    mPathToConFile = QDir::currentPath();
+    mPathToConFile += "/conf/";
+    mPathToConFile += mConFile;
+}
+
+void StoreConf::checkConfDir(QString str)
+{
+    if (str == "exe")
+    {
```
Comment (vsakharevych) [inline src/settings/storeconf.cpp:79]: current method is called only once in code (line 31) & is not passed any arg
Therefore default param will be used in 100% cases
Than this if statement will evaluate to true in any case

File: src/settings/storeconf.cpp [hunk 2]
```diff
@@ -9,126 +9,165 @@
...
-    loadDoc = QJsonDocument::fromJson(saveData);
-    if(loadDoc.isNull() || !loadDoc.isObject() || loadDoc.isEmpty())
-        readStatus = false;
+    mJsonDoc = QJsonDocument::fromJson(saveData);
+    if(mJsonDoc.isNull() || !mJsonDoc.isObject() || mJsonDoc.isEmpty())
+        mReadStatus = false;
     else
-        readStatus = true;
+        mReadStatus = true;
 }
```
Comment (vsakharevych) [inline src/settings/storeconf.cpp:197]: Value of mReadStatus is overwritten several times without need
There is only one case when it should be assigned true

File: src/settings/storeconf.cpp [hunk 3]
```diff
@@ -9,126 +9,165 @@
...
+        if (mReadStatus)
+        {
+            parseJson();
+        }
+        else
+        {                        //  file corrupted
+            writeJson();        //      rewrite with default values
+        }
         saveData();
     }
```
Comment (vsakharevych) [inline src/settings/storeconf.cpp:55]: saveDate can be moved from if / else statement to avoid duplication

File: src/settings/storeconf.cpp [hunk 4]
```diff
@@ -9,126 +9,165 @@
...
+    {
+        if (mode == "start")
+        {                           // copy value from QMap to json
+            root_obj.insert(it.key(), it.value());
+        }
+        else
+        {                           // copy value from QSettings to json
+            if (s.contains(it.key()))
+                root_obj.insert(it.key(), s.value(it.key()).toString());
+        }
```
Comment (vsakharevych) [inline src/settings/storeconf.cpp:162]: Why are data written to both QSettings repository & json file?

File: src/settings/storeconf.h [hunk 5]
```diff
@@ -14,37 +14,39 @@
...
+    QMap<QString, QString> mFields; //  first -> valueName, second -> valueItself
+
+    QStringList mCppExtentionsList; //  all extentions available in C++ mode
+
+    void checkConfDir(QString str = "exe");    //  check existance directory <conf>, create if not exist
+                                               //  str = "exe" - directory in same dir as execution file
+    void writeJson(QString str = "start");     //  on application start: if conf.json not exist - creates it in conf directory
+                                               //  on application quit:  rewrite json with values from QSettings
+    void readJson();                //  read configuration file to [loadDoc]. status write to [readStatus]
     void parseJson();
```
Comment (vsakharevych) [inline src/settings/storeconf.h:49]: readJson & writeJson both indicate execution status using flags mReadStatus/mWriteStatus
To my opinion it is better to use exceptions (C++ approach) or at least return bool (C approach)

Also I don't see if you use mWriteStatus at all

File: src/settings/storeconf.cpp (hunk 1 above)
Comment (MarsLviv) [reply src/settings/storeconf.cpp:79]: to get posibility to change location directory for configuration files signature of checkConDir has argument which passes this information
by default it lives in same directory as execution file: so parameter = "exe"

File: src/settings/storeconf.cpp (hunk 2 above)
Comment (MarsLviv) [reply src/settings/storeconf.cpp:197]: reading of json file can fall in two ways: during  it opening as raw data (line 118) and during it transforming to type QJsonDocument (line 129)
so it checked two times

File: src/settings/storeconf.h (hunk 5 above)
Comment (MarsLviv) [reply src/settings/storeconf.h:49]: ok
I remove mWriteStatus

File: src/settings/storeconf.cpp (hunk 3 above)
Comment (MarsLviv) [reply src/settings/storeconf.cpp:55]: done

File: src/settings/storeconf.cpp (hunk 4 above)
Comment (MarsLviv) [reply src/settings/storeconf.cpp:162]: all values from configuration file application load during starting and save values at the quiting.
during current session all changes (can be many times) reflecting in QSettings as global available place.

**Review by vsakharevych — APPROVED** [review]

File: src/editor/codeeditor.cpp [hunk 6]
```diff
@@ -20,9 +20,9 @@ CodeEditor::CodeEditor(QWidget *parent) : QPlainTextEdit(parent)
 
     //read settings
     QSettings settings(QApplication::organizationName(), QApplication::applicationName());
-    QString analizerFontSize = settings.value("analizerFontSize").toString();
-    QString analizerFontName = settings.value("analizerFontName").toString();
-    QString analizerStyle = settings.value("analizerStyle").toString();
+    QString analizerFontSize = settings.value("editorFontSize").toString();
```
Comment (valentynfaychuk) [inline src/editor/codeeditor.cpp:23]: analyzer...

File: src/settings/storeconf.cpp [hunk 7]
```diff
@@ -9,126 +9,165 @@
...
-    {                   // create conf.json with these values
+    mFields["applicationName"] = "Pair Storm";       //  default values used when json file
+    mFields["applicationVersion"] = "0.1";           //      not exist or corrupted
+    mFields["organizationName"] = "Lv-425.C++";
+    mFields["style"] = "WHITE";
+    mFields["editorFontName"] = "Consolas";
+    mFields["editorFontSize"] = "12";
+    mFields["defaultUdpPortNumber"] = "36108";
+    mFields["defaultTcpPortNumber"] = "32807";
+    mFields["userName"] = "unnamed user";
```
Comment (valentynfaychuk) [inline src/settings/storeconf.cpp]: Why do you pass "unnamed user" to **mFields[**"userName"**]** when you have **userName** parameter.
Think about typing in parameter list **QString userName = "unnamed user"** and in the code **mFields[**"userName"**] = userName;**

File: src/settings/storeconf.cpp [hunk 8]
```diff
@@ -9,126 +9,165 @@
...
+    {
+        QString pathToConDir = QDir::currentPath();
+        pathToConDir += "/conf";
+        if (!QDir(pathToConDir).exists())
+            QDir().mkdir(pathToConDir);
+    }
 }
 
-void StoreConf::writeJson()
+void StoreConf::writeJson(QString mode)
```
Comment (valentynfaychuk) [inline src/settings/storeconf.cpp]: I recommend you changing **mode** from QString to enumeration

File: src/settings/storeconf.cpp [hunk 9]
```diff
@@ -9,126 +9,165 @@
...
-    QApplication::setApplicationName(applicationName);
+    QApplication::setOrganizationName(mFields["organizationName"]);
+    QApplication::setApplicationVersion(mFields["applicationVersion"]);
+    QApplication::setApplicationName(mFields["applicationName"]);
+
+    QSettings settings;
+    QMap<QString, QString>::const_iterator it = mFields.constBegin();
+    while (it != mFields.constEnd())
+    {
+        settings.setValue(it.key(), mFields[it.key()]);        // copy values from QMap to QSettings
```
Comment (valentynfaychuk) [inline src/settings/storeconf.cpp:250]: What if there is already saved value?

**Review by valentynfaychuk — CHANGES_REQUESTED** [review]
Reduce amount of flags inside your class as Volodya said. This should reduce coupling and decrease complexity.

File: src/editor/codeeditor.cpp (hunk 6 above)
Comment (MarsLviv) [reply src/editor/codeeditor.cpp:23]: I changed key names that better represent it sence.
all new key names in storeconf.cpp

File: src/settings/storeconf.cpp (hunk 7 above)
Comment (MarsLviv) [reply src/settings/storeconf.cpp]: userName is a key
key should be common for all cases

File: src/settings/storeconf.cpp (hunk 9 above)
Comment (MarsLviv) [reply src/settings/storeconf.cpp:250]: in start session user read value from settings
during session can change it couple times
method setValue() rewrite value

File: src/settings/storeconf.cpp (hunk 8 above)
Comment (MarsLviv) [reply src/settings/storeconf.cpp]: changed to enumeration

File: src/settings/storeconf.cpp (hunk 7 above)
Comment (MarsLviv) [reply src/settings/storeconf.cpp]: changed to userName

**Review by valentynfaychuk — APPROVED** [review]

**Pull request merged**